src/rust/catmem/queue.rs

```diff
@@ -94,71 +94,48 @@ impl CatmemQueue {
         }
     }
 
-    /// This function closes a ring endpoint.
-    /// TODO merge this with async_close().
-    pub fn close(&self) -> Result<(), Fail> {
-        match &mut *self.ring.borrow_mut() {
-            Ring::PushOnly(ring) => {
-                // Attempt to push EoF.
-                // Maximum number of retries. This is set to an arbitrary small value.
-                for _ in 0..MAX_RETRIES_PUSH_EOF {
-                    match ring.try_close() {
-                        Ok(()) => return Ok(()),
-                        Err(_) => continue,
-                    }
-                }
-                let cause: String = format!("failed to push EoF");
-                error!("push_eof(): {}", cause);
-                Err(Fail::new(libc::EIO, &cause))
-            },
-            // Nothing to do on close for a pop ring.
-            Ring::PopOnly(_) => Ok(()),
-        }
-    }
-
-    /// Prepares a transition to the closing state.
-    pub fn prepare_close(&self) -> Result<(), Fail> {
-        match &mut *self.ring.borrow_mut() {
-            Ring::PushOnly(ring) => ring.prepare_close(),
-            Ring::PopOnly(ring) => ring.prepare_close(),
-        }
-    }
-
-    /// Prepares a transition to the closed state.
-    pub fn prepare_closed(&self) -> Result<(), Fail> {
-        match &mut *self.ring.borrow_mut() {
-            Ring::PushOnly(ring) => ring.prepare_closed(),
-            Ring::PopOnly(ring) => ring.prepare_closed(),
+    pub fn shutdown(&mut self) -> Result<(), Fail> {
+        {
+            let mut ring: RefMut<Ring> = self.ring.borrow_mut();
+            ring.prepare_close()?;
+            ring.commit();
+            ring.prepare_closed()?;
+            ring.commit();
         }
-    }
+        self.cancel_pending_ops(Fail::new(libc::ECANCELED, "this queue was shutdown"));
 
-    /// This function commits the queue to closing.
-    pub fn commit(&self) {
-        match &mut *self.ring.borrow_mut() {
-            Ring::PushOnly(ring) => ring.commit(),
-            Ring::PopOnly(ring) => ring.commit(),
-        }
+        Ok(())
     }
 
-    /// This function aborts a pending operation.
-    pub fn abort(&self) {
-        match &mut *self.ring.borrow_mut() {
-            Ring::PushOnly(ring) => ring.abort(),
-            Ring::PopOnly(_) => warn!("abort() called on a pop-only queue"),
+    /// This function closes a ring endpoint.
+    /// TODO merge this with async_close().
+    pub fn close(&mut self) -> Result<(), Fail> {
```

Comment:
Shutdown and close are identical, perhaps one of them one of them could just call the other?

Reply:
Shutdown doesn't call close, it just sets the queue to the Closed state and cancels pending ops.

Reply:
Ah ok, yes, I misread as they are very similar :). Please ignore my comment.